File: Watkins_Q_learning.f90
module Watkins_Q_learning
  use func
  implicit none

contains

  !------------------------------------------------------------
  ! This function defines the ramp of RL inverse temperature
  ! input
  ! ts : episode number/time
  ! m  : slope of increase
  ! b  : y intercept
  ! Tl : duration of ramp before zeroing
  !------------------------------------------------------------
  function explore_beta(ts, m, b, Tl, beta_RL_const)
    implicit none
    integer,intent(in)::ts, Tl
    real*8,intent(in)::m, b, beta_RL_const
    real*8 explore_beta

    if(mod((int(ts/Tl)), 2)==1) then
      explore_beta = beta_RL_const
    else
      explore_beta = dble(b+m/2.d0*(dble(1.d0*ts/Tl)-int(ts/Tl)/2.d0))
    end if

  end function explore_beta

  !----------------------------------------------------------
  ! This function finds the feature indices of state S
  !----------------------------------------------------------
  subroutine find_feature_inds(tilings, S, theta_inds)
    implicit none
    real*8,intent(in),dimension(:,:)::tilings
    real*8,intent(in)::S
    integer,intent(inout),dimension(:)::theta_inds
    integer n1, n2, i, j, k, idx
    real*8 left, right

    n1 = size(tilings, 1)
    n2 = size(tilings, 2)
    do i = 1, n1
      call search_sorted(tilings(i, :), S, idx)
      call clip(1, n2-1, idx)
      left = tilings(i, idx)
      right = tilings(i, idx+1)
      if (S-left>right-S) then
        idx = idx+1
      end if
      theta_inds(i) = idx +(i-1)*n2
    end do

  end subroutine find_feature_inds

  !--------------------------------------------------------------
  ! This function biases a given Q function (stored in theta)
  ! to learn a policy by scaling the weights of all actions
  ! down below the value of the best actions
  !--------------------------------------------------------------
  subroutine Learn_Policy(state_i, best_actions, R, theta, tilings, actions)
    implicit none
    real*8,intent(in)::state_i
    real*8,intent(in),dimension(:)::best_actions, actions
    real*8,intent(in)::R
    real*8,intent(inout),dimension(:,:,:)::theta
    real*8,intent(in),dimension(:,:)::tilings
    ! temporary variables
    integer N_tilings, N_tiles, N_actions
    integer t_step, indA, indA_max(1)
    real*8 A, S
    integer,allocatable,dimension(:)::theta_inds, Q

    N_tilings = size(tilings, 1)
    N_tiles = size(tilings, 2)
    N_actions = size(actions)
    allocate(theta_inds(N_tilings))
    allocate(Q(N_actions))
    S = state_i

    do t_step = 1, size(best_actions)
      ! SA index
      A = best_actions(t_step)
      call search(actions, A, indA)
      call find_feature_inds(tilings, S, theta_inds)

      ! check if max learnt
      if (maxval(theta(theta_inds, t_step, :))>dble(R/N_tilings) .and. R>0.d0) then
        Q = sum(theta(theta_inds, t_step, :), 1)
        indA_max = maxloc(Q)

        if (maxval(Q)>1.d-13) then
          theta(theta_inds, t_step, :) = theta(theta_inds, t_step, :)*R/Q(indA_max(1))
        end if
      end if

      ! calculate theta function
      theta(theta_inds, t_step, indA) = dble((R+1.d-2)/N_tilings)
      S = S+A
    end do

    deallocate(theta_inds)
    deallocate(Q)

  end subroutine Learn_Policy

  !-----------------------------------------------------
  ! This function builds a protocol from actions set
  !-----------------------------------------------------
  subroutine build_protocol(actions_taken, q_i, delta_t, protocol, t_ls)
    implicit none
    real*8,intent(in),dimension(:)::actions_taken
    real*8,intent(in)::q_i, delta_t
    real*8,intent(inout),dimension(:)::protocol, t_ls
    integer i, j, k
    real*8 A, S

    S = q_i
    do i = 1, size(actions_taken)
      A = actions_taken(i)
      S = S+A
      protocol(i) = S
      t_ls(i) = (i-1)*delta_t
    end do

  end subroutine build_protocol

  !--------------------------------------------------------
  ! This function gives the reward (fidelity on TF state)
  !--------------------------------------------------------
  function fidelity(psi)
    implicit none
    complex*16,intent(in),dimension(:)::psi
    real*8 fidelity

    fidelity = abs(psi(size(psi)))**2

  end function fidelity

  !---------------------------------------------------------
  ! Modified true on-line Watkins's Q-learning algorithm
  !---------------------------------------------------------
  subroutine Q_learning(Ntot, N_realize, N_episodes, alpha_0, eta, lmbda, beta_RL_i, beta_RL_inf, &
                      & T_expl, m_expl, N_tilings, N_tiles, state_i, q_field, dq_field, &
                      & max_t_steps, delta_time, q_i, q_f, psi_i, psi_f, expm_dict)
    implicit none
    ! input & output parameters
    integer,intent(in)::Ntot, N_realize, N_episodes, T_expl, N_tilings, N_tiles, max_t_steps
    real*8,intent(in)::alpha_0, eta, beta_RL_i, beta_RL_inf, m_expl, state_i, delta_time
    real*8,intent(in)::q_i, q_f, lmbda, dq_field
    real*8,intent(in),dimension(:)::q_field
    complex*16,intent(in),dimension(:)::psi_i, psi_f
    complex*16,intent(inout),dimension(:,:,:)::expm_dict
    ! temporary variables
    real*8,allocatable,dimension(:,:,:)::theta
    real*8,allocatable,dimension(:,:)::tilings
    real*8,allocatable,dimension(:)::fire_trace, u0, u, Q, actions_taken, avail_actions, &
                                   & best_actions, alpha, protocol, t_ls
    complex*16,allocatable,dimension(:)::psi
    integer,allocatable,dimension(:)::theta_inds, avail_inds, theta_inds_prime
    real*8,allocatable,dimension(:,:,:)::theta_old, e
    real*8,dimension(5)::actions
    integer,dimension(5)::avail_inds_tmp
    integer N_actions, cnt, ep, t_step, avail_inds_cnt, indA
    real*8 q_shift, best_R, R, S, S_prime, beta_RL, A_greedy, A, delta_t, &
         & Q_old, delta_q

    ! preallocate physical state and define actions
    allocate(psi(size(psi_i)))
    psi = cmplx(0.d0, 0.d0)
    actions = [-12.d0, -6.d0, 0.d0, 6.d0, 12.d0]
    N_actions = size(actions)

    ! preallocate theta and tilings
    allocate(theta(N_tiles*N_tilings, max_t_steps, N_actions))
    allocate(tilings(N_tilings, N_tiles))

    ! preallocate weight vector theta and tilings
    theta = 0.d0
    allocate(theta_old(size(theta,1),size(theta,2),size(theta,3)))
    theta_old = theta
    do cnt = 1, N_tilings
      call random_number(q_shift)
      tilings(cnt, :) = q_field+q_shift*dq_field
    end do

    ! preallocate eligibility trace
    allocate(e(size(theta,1),size(theta,2),size(theta,3)))
    e = 0.d0
    allocate(fire_trace(N_tilings))
    fire_trace = 1.d0

    ! preallocate usage vector : inverse gradient descent learning rate
    allocate(u0(N_tiles*N_tilings))
    allocate(u(N_tiles*N_tilings))
    u = 0.d0
    u0 = 1.d0/alpha_0

    ! preallocate reward and feature index
    best_R = -1.d0
    R = 0.d0
    allocate(theta_inds(N_tilings))
    allocate(theta_inds_prime(N_tilings))
    theta_inds = 0

    ! preallocate Q function
    allocate(Q(N_actions))
    Q = 0.d0

    ! preallocate action taken list
    allocate(actions_taken(max_t_steps))

    ! preallocate best action
    allocate(best_actions(max_t_steps))

    ! preallocate alpha
    allocate(alpha(N_tilings))

    !-------------------------------
    !  Loop over multiple episodes
    !-------------------------------
    do ep = 1, N_episodes

      e = 0.d0 ! init eligibility trace to zero
      u = u0 ! init learning rate
      S = state_i ! init state S at t = 0

      ! get feature of S
      call find_feature_inds(tilings, S, theta_inds)

      ! init Q(S,t=0,:) for all actions
      Q = sum(theta(theta_inds, 1, :), 1)

      ! preallocate physical state
      psi = psi_i

      ! init taken actions
      actions_taken = 0.d0

      ! define learning temperature
      beta_RL = explore_beta(ep-1, m_expl, beta_RL_i, T_expl, beta_RL_inf)

      !---------------------------------------
      !   Loop over time step in episide ep
      !---------------------------------------
      do t_step = 1, max_t_steps

        ! available actions for state S at time t = t_step
        avail_inds_cnt = 0
        do cnt = 1, N_actions
          if (S+actions(cnt)<=q_field(size(q_field)) .and. S+actions(cnt)>=q_field(1)) then
            avail_inds_cnt = avail_inds_cnt+1
            avail_inds_tmp(avail_inds_cnt) = cnt
          end if
        end do
        allocate(avail_inds(avail_inds_cnt))
        allocate(avail_actions(avail_inds_cnt))
        avail_inds(:) = avail_inds_tmp(1:avail_inds_cnt)
        avail_actions = actions(avail_inds)

        ! determine greedy action for S at time t = t_step
        if (beta_RL<beta_RL_inf) then
          if (mod(ep-1, 2)==0) then
            A_greedy = avail_actions(random_choice_from_max(Q(avail_inds)))
          else
            A_greedy = best_actions(t_step)
          end if
        else
          A_greedy = avail_actions(random_choice_from_max(Q(avail_inds)))
        end if

        ! choose action for S at time t = t_step by explore & exploit phase
        if (beta_RL<beta_RL_inf) then ! explore by Boltzman-policy
          A = avail_actions(random_choice_from_BP(exp(beta_RL*Q(avail_inds))))

          ! reset eligibility trace to 0 if A is exploratory
          if (abs(A-A_greedy)>1.d-12) then
            e = 0.d0
          end if
        else ! exploit
          A = A_greedy
        end if

        ! index of A in actions
        call search(actions, A, indA)

        ! record action taken
        actions_taken(t_step) = A

        !------------------------------------
        ! determine reward by time evolution
        !------------------------------------
        S_prime = S
        S_prime = S_prime+A ! next state S' at t = t_step+1
        ! physical state at t = t_step+1
        psi = matmul(expm_dict(nint((S_prime-minval(q_field))/0.1d0)+1, :, :), psi)
        call normalize(psi)

        ! asign reward
        R = 0.d0 ! non-terminal state
        if (t_step==max_t_steps) then ! terminal state
          R = R+fidelity(psi)
        end if

        ! learning rate update
        u(theta_inds) = (1.d0-eta)*u(theta_inds)
        u(theta_inds) = u(theta_inds)+1.d0
        alpha = 1.d0/(N_tilings*u(theta_inds))

        !--------------------------------------
        ! Q-learning TD error and update rule
        !--------------------------------------
        delta_t = R-Q(indA) ! R(t+1)-Q(St,A)
        Q_old = sum(theta(theta_inds, t_step, indA)) ! Q(S,t,A)
        e(theta_inds, t_step, indA) = alpha*fire_trace ! e(S,t,A)<-1

        ! check if S_prime is terminal or out of range
        if (t_step==max_t_steps) then
          ! update theta
          theta = theta+delta_t*e
          ! GD error in field q
          delta_q = Q_old-sum(theta(theta_inds, t_step, indA))
          theta(theta_inds, t_step, indA) = theta(theta_inds, t_step, indA) &
                                          & +delta_q*alpha
          ! deallocate temporary variables inside loop
          deallocate(avail_actions)
          deallocate(avail_inds)
          ! go to next episode
          exit
        end if

        ! feature of S_prime
        call find_feature_inds(tilings, S_prime, theta_inds_prime)

        ! t-dependent Watkins's Q-learning
        Q = sum(theta(theta_inds_prime, t_step+1, :), 1) ! Q(S',t+1,:)

        ! update theta
        delta_t = delta_t+maxval(Q) ! max_a[Q(S',t+1,a)] for all a
        theta = theta+delta_t*e

        ! TD error in field q
        delta_q = Q_old-sum(theta(theta_inds, t_step, indA))
        theta(theta_inds, t_step, indA) = theta(theta_inds, t_step, indA) &
                                        & +delta_q*alpha

        ! update trace e(S,t,A) and e(:,:,:)
        e(theta_inds, t_step, indA) = e(theta_inds, t_step, indA) &
                                    & -sum(e(theta_inds, t_step, indA))*alpha
        e = lmbda*e

        ! update state and its feature
        S = S_prime
        theta_inds = theta_inds_prime

        ! deallocate temporary variables inside loop
        deallocate(avail_actions)
        deallocate(avail_inds)

      end do

      ! best policy so far
      if (R-best_R>1.d-12) then
        ! write(*, *) "best fidelity is : ", R, "with episodes : ", ep
        ! update list best action so far
        best_actions = actions_taken
        ! update best reward and weight vector theta
        best_R = R
        call Learn_Policy(state_i, best_actions, best_R, theta, tilings, actions)
      end if

      ! force-learn replay every 100 episides
      if ((mod(ep+1, 2*T_expl)-T_expl==0) .and. (ep .ne. 1) .and. (ep .ne. N_episodes)) then
        call Learn_Policy(state_i, best_actions, best_R, theta, tilings, actions)
      elseif ((mod(int(ep/T_expl), 2)==1) .and. (abs(R-best_R)>1.d-12)) then
        call Learn_Policy(state_i, best_actions, best_R, theta, tilings, actions)
      end if

      ! check convergence of theta function
      ! write(*, '(I4,3F12.6)') ep, beta_RL, R, maxval(abs(theta-theta_old))
      theta_old = theta

    end do

    allocate(protocol(max_t_steps))
    allocate(t_ls(max_t_steps))
    call build_protocol(actions_taken, state_i, delta_time, protocol, t_ls)
    ! write(*, *) protocol

    ! deallocate temporary variables outside loop
    deallocate(psi)
    deallocate(theta)
    deallocate(tilings)
    deallocate(theta_old)
    deallocate(e)
    deallocate(fire_trace)
    deallocate(u0)
    deallocate(u)
    deallocate(theta_inds)
    deallocate(theta_inds_prime)
    deallocate(Q)
    deallocate(actions_taken)
    deallocate(best_actions)
    deallocate(alpha)
    deallocate(protocol)
    deallocate(t_ls)

  end subroutine Q_learning


end module Watkins_Q_learning

File: Q_learning_python3/Watkins_Q_learning.py
from ED import *
import numpy as np
import numpy.random as random
import time
import os
import sys
import pickle
import cPickle

def explore_beta(t, m, b, T, beta_RL_const = 1000.0):
	"""
	This function defines the ramp of the RL inverse temperature:

	t: episode number/time
	m: slope of increase
	b: y intercept
	T: duration of ramp before zeroing
	"""
	if (t//T)%2==1:
		return beta_RL_const
	else:
		return b + m/2.0*(float(t)/T - (t//T)/2.0)


def find_feature_inds(tilings, S, theta_inds):
	"""
	This function finds the feature indices of s state S in the set of tilings.
	"""
	for _k, tiling in enumerate(tilings): # cython-ise this loop as inline fn!
		idx = tiling.searchsorted(S)
		idx = np.clip(idx, 1, len(tiling)-1)
		left, right = tiling[idx-1], tiling[idx]
		idx -= S - left < right - S
		theta_inds[_k]=idx[0]+_k*tilings.shape[1]
	return theta_inds


def Learn_Policy(state_i, best_actions, R, theta, tilings, actions):
	"""
	This function biases a given Q function (stored in the linear approximant theta)
	to learn a policy by scaling the weights of all actions down below the value of the
	best action.
	"""
	N_tilings = tilings.shape[0]
	N_tiles = tilings.shape[1]
	# preallocate theta_inds
	theta_inds_zeros = np.zeros((N_tilings,), dtype=int)
	# init state at t=0
	S = state_i.copy()

	for t_step, A in enumerate(best_actions):
		# SA index
		indA = np.searchsorted(actions, A)
		theta_inds = find_feature_inds(tilings, S, theta_inds_zeros)

		# check if max learnt
		if max(theta[theta_inds,t_step,:].ravel())>R/N_tilings and R>0: # weights
			Q = np.sum(theta[theta_inds,t_step,:],axis=0)
			indA_max=np.argmax(Q)

			if max(Q) > 1E-13:
				theta[theta_inds,t_step,:]*=R/Q[indA_max]

		# calculate theta function
		theta[theta_inds,t_step,indA] = (R+1E-2)/(N_tilings)

		S+=A

	#print 'force-learned best encountered policy'
	return theta


def build_protocol(actions, q_i, delta_t):
	""" This function builds a protocol from the set of actions """
	protocol = np.zeros_like(actions)
	t = np.array([delta_t*_i for _i in range(len(actions))])
	S = q_i
	for _i,A in enumerate(actions):
		S += A
		protocol[_i] = S
	return protocol, t


def greedy_protocol(theta, tilings, actions, q_i, delta_t, max_t_steps, q_field):
	""" This function builds the best encounteres protocol from best_actions """
	protocol = np.zeros((max_t_steps,), dtype=np.float64)
	t = np.array([delta_t*_i for _i in range(max_t_steps)])
	S = np.array([q_i])

	N_tilings = tilings.shape[0]
	N_tiles = tilings.shape[1]

	# preallocate theta_inds
	theta_inds_zeros = np.zeros((N_tilings,), dtype=int)

	for t_step in range(max_t_steps):

		avail_inds = np.argwhere((S[0]+np.array(actions)<=q_field[-1])*(S[0]+np.array(actions)>=q_field[0])).squeeze()
		avail_actions = actions[avail_inds]

		# calculate Q(s,a)
		theta_inds = find_feature_inds(tilings,S,theta_inds_zeros)
		Q = np.sum(theta[theta_inds,t_step,:],axis=0)
		# find greedy action
		A_greedy = avail_actions[np.argmax(Q[avail_inds])]

		S[0] += A_greedy
		protocol[t_step] = S

	return protocol, t


def reward(psi):
	""" This function gives the reward according to spin population """
	Ntot = int((len(psi)-1)*2)
	dim = int(Ntot/2)+1
	ls = np.array([i*1.0 for i in range(dim)], dtype=np.float64)
	rw = 2.0*np.sum(ls*(abs(psi)**2))/Ntot

	return rw

def fidelity(psi):
	""" This function gives the reward according to fidelity on twin-Fock state  """
	rw = abs(psi[-1])**2

	return rw

def weight_fid(psi):
	"""
	This function gives the reward according to the wighted fidelity on the final
	ten percents of basis satate around the twin-Fock state
	"""
	dim = len(psi)
	gamma = 0.9 # discount rate
	num = 100 # the last 100 basis states are considered
	rw = 0.0
	for cnt, i in enumerate(range(dim-num, dim)):
		rw += (gamma**(num-1-cnt))*(abs(psi[i])**2)
	rw = (1.0-gamma)*rw

	return rw

#----------------------------------------------------------
#	Modified true on-line Watkins's Q(lambda) algorithm
#----------------------------------------------------------
def Q_learning(Ntot, N,N_episodes,alpha_0,eta,lmbda,beta_RL_i,beta_RL_inf,T_expl,m_expl,N_tilings,N_tiles,state_i,q_field,dq_field,
			   max_t_steps,delta_time,q_i,q_f,psi_i,psi_f,expm_dict, file1,file2,
			   theta=None,tilings=None,save=False):
	"""
	This function applies modified Watkins' Q-Learning for time-dependent states with
	force-learn replays.

	1st row: RL arguments
	2nd row: physics arguments
	3rd row: optional arguments
	"""

	# global variable

	# preallocate unitary operator list
	# expm_dict = cPickle.load(open('input/expm_dict_'+str(N)+'_'+str(Ntot)+'.pkl', 'rb'))

	# preallocate physical state
	psi = np.zeros_like(psi_i)
	# define actions
	pos_actions = [0.1, 0.2, 0.5, 1.0, 2.0, 4.0, 6.0]
	neg_actions = [-i for i in pos_actions]
	actions = np.sort(neg_actions+[0.0]+pos_actions)
	N_actions = len(actions)

	# preallocate unitary operator list

	# preallocate weight vector theta
	if theta is None:
		theta=np.zeros((N_tiles*N_tilings, max_t_steps, N_actions), dtype=np.float64)
	theta_old=theta.copy()
	if tilings is None:
		tilings = np.array([q_field+np.random.uniform(0.0,dq_field,1) for j in xrange(N_tilings)])

	# preallocate eligibility trace
	e = np.zeros_like(theta) # for all state, time, actions
	fire_trace = np.ones(N_tilings) # for update SAP

	# pre-allocate usage vector: inverse gradient descent learning rate
	u0 = 1.0/alpha_0*np.ones((N_tiles*N_tilings,), dtype=np.float64)
	u=np.zeros_like(u0)

	# preallocate quantities
	Return_ave = np.zeros((N_episodes,),dtype=np.float64)
	Return = np.zeros_like(Return_ave)
	Fidelity_ep = np.zeros_like(Return_ave)
	protocol_ep = np.zeros((Fidelity_ep.shape[0],max_t_steps),)

	best_R = -1.0 # init best reward
	R = 0.0 # init reward
	theta_inds_zeros = np.zeros((N_tilings,), dtype=int)

	#-----------------------------------
	#	Loop over multiple episodes
	#-----------------------------------
	for ep in xrange(N_episodes):

		e *= 0.0 # init eligibility trace to zero
		u[:] = u0[:] # init learning rate
		S = state_i.copy() # init state S at t=0

		# get feature of S
		theta_inds = find_feature_inds(tilings, S, theta_inds_zeros)

		# init Q(S,t=0,:) for all actions
		Q = np.sum(theta[theta_inds, 0, :], axis = 0)

		# preallocate physical state
		psi[:] = psi_i[:]

		# store taken actions
		actions_taken = np.zeros((max_t_steps,), dtype=np.float64)

		# define learning temperature
		beta_RL = explore_beta(ep,m_expl,beta_RL_i,T_expl,beta_RL_const=beta_RL_inf)

		# explore status
		explored = False

		#---------------------------------------
		#	Loop over time step in episode ep
		#---------------------------------------
		for t_step in xrange(max_t_steps):

			# avaliable actions for state S at time t=t_step
			avail_inds = np.argwhere((S[0]+np.array(actions)<=q_field[-1])*(S[0]+np.array(actions)>=q_field[0])).squeeze()
			avail_actions = actions[avail_inds]

			# determine greedy action for S at time t=t_step
			if beta_RL<beta_RL_inf:
				if ep%2==0:
					A_greedy = avail_actions[random.choice(np.argwhere(Q[avail_inds]==np.amax(Q[avail_inds])).ravel() ) ]
				else:
					A_greedy = best_actions[t_step]
			else:
				A_greedy = avail_actions[random.choice(np.argwhere(Q[avail_inds]==np.amax(Q[avail_inds])).ravel() ) ]

			# choose action for S at time t=t_step by explore & exploit phase
			if beta_RL<beta_RL_inf: # explore by Boltzman-policy
				P = np.exp(beta_RL*Q[avail_inds])
				A = avail_actions[np.searchsorted(np.cumsum(P/np.sum(P)),random.uniform(0.0,1.0))]

				# reset eligibility trace to 0 if A is exploratory
				if abs(A-A_greedy)>np.finfo(A).eps:
					e *= 0.0
			else: # exploit
				A = A_greedy

			# index of A in actions
			indA = np.searchsorted(actions, A)

			# record action taken
			actions_taken[t_step] = A

			#-----------------------------------------
			#	determine the reward by evolution
			#-----------------------------------------
			S_prime = S.copy()
			S_prime[0] += A # next state S' at t=t_step+1
			# physical state at t=t_step+1
			b = S_prime[0]
			#psi = exp_H.dot(psi)
			psi=expm_dict[int(np.rint((b-min(q_field))/min(pos_actions)))].dot(psi)

			# asign reward
			R *= 0.0 # non-terminal state
			if t_step==max_t_steps-1: # terminal state
				# R += fidelity
				# R += abs(psi.conj().dot(psi_f))**2
				R += fidelity(psi)
				# R += fidelity(psi)
				# R += weight_fid(psi)

			#----------------------------------------
			#			learning rate update
			#----------------------------------------
			u[theta_inds] *= (1.0-eta)
			u[theta_inds] += 1.0
			alpha = 1.0/(N_tilings*u[theta_inds]) # update learning rate

			#-----------------------------------
			# Q(lambda) TD error and update rule
			#-----------------------------------
			delta_t = R-Q[indA] # R(t+1)-Q(St,A)
			Q_old = theta[theta_inds, t_step, indA].sum() # Q(S,t,A)
			e[theta_inds, t_step, indA] = alpha*fire_trace # e(S,t,A)<-1

			# check if S_prime is terminal or out of range
			if t_step==max_t_steps-1:
				# update theta
				theta += delta_t*e
				# GD error in field q
				delta_q = Q_old - theta[theta_inds, t_step, indA].sum()
				theta[theta_inds, t_step, indA] += alpha*delta_q
				# go to next episode
				break

			# feature of S_prime
			theta_inds_prime=find_feature_inds(tilings, S_prime, theta_inds_zeros)

			# t-dependent Watkins's Q-learning
			Q = np.sum(theta[theta_inds_prime,t_step+1,:],axis=0) # Q(S',t+1,:)

			# update theta
			delta_t += np.max(Q) # max_a(Q(S',t+1,a)) for all a
			theta += delta_t*e

			# TD error in field q
			delta_q = Q_old - theta[theta_inds, t_step, indA].sum()
			theta[theta_inds, t_step, indA] += alpha*delta_q

			# update traces e(S,t,A) and e(:,:,:)
			e[theta_inds, t_step, indA] -= alpha*e[theta_inds, t_step, indA].sum()
			e *= lmbda

			# update state and its feature
			S[:] = S_prime[:]
			theta_inds[:] = theta_inds_prime[:]

		# best policy so far
		if R-best_R>1E-12:
			print("best encountered fidelity is {} with training number {}".format(np.around(R, 4), ep))
			# update list of best actions so far
			best_actions = actions_taken[:]
			# update best reward and weight vector theta
			best_R = R
			theta = Learn_Policy(state_i, best_actions, best_R, theta, tilings, actions)

		# force-learn replay every 100 episodes
		if ((ep+1)%(2*T_expl)-T_expl==0 and ep not in [0,N_episodes-1]): # and beta_RL<20.0:
			theta = Learn_Policy(state_i, best_actions, best_R, theta, tilings, actions)
		elif (ep//T_expl)%2==1 and abs(R-best_R)>1E-12:
			theta = Learn_Policy(state_i, best_actions, best_R, theta, tilings, actions)

		# # check convergence of Q-function
		# print ep, "beta_RL,R,d_theta:", beta_RL, R, np.max(abs(theta.ravel()-theta_old.ravel()))
		# theta_old=theta.copy()

		# # record quantities
		# Return_ave[ep] = 1.0/(ep+1)*(R+ep*Return_ave[ep-1])
		# Return[ep] = R
		# Fidelity_ep[ep] = R
		# protocol_ep[ep,:] = build_protocol(actions_taken, state_i[0], delta_time)[0].astype(int)

		# if (ep+1)%(2*T_expl)==0:
		# 	print "finished simulating episode {} with fidelity {} at hx_f = {}.".format(ep+1,np.round(R,5),S_prime[0])
		# 	#print 'best encountered fidelity is {}.'.format(np.round(best_R,5))
		# 	#print 'current inverse exploration tampeature is {}.'.format(np.round(beta_RL,3))

		# if (ep%100==0): # ouput every 400 episodes
		# 	# print('episodes {} with reward {}'.format(ep, best_R))
		# 	cPickle.dump(best_R, file2)

	# # determine the global best protocol and fidelity
	# protocol_best, t_best = build_protocol(best_actions, state_i[0], delta_time)
	# protocol_greedy, t_greedy = greedy_protocol(theta, tilings, actions, state_i[0], delta_time, max_t_steps, q_field)
    # 
	# out_data = {}
	# out_data[0] = max_t_steps*delta_time
	# out_data[1] = best_R
	# out_data[2] = protocol_best
	# cPickle.dump(out_data, file1)
	# # cPickle.dump(best_R, file2)
	# print('best reward is {} with {} episodes'.format(best_R, N_episodes-1))








